File: crates/lillibrain/src/freelist.rs
//! Trunk-page freelist over the pager.
//!
//! Freed pages are recorded on a chain of trunk pages so they can be reused
//! before the file is extended. A trunk page stores a pointer to the next
//! trunk, a count of free leaf page numbers, and the leaf page numbers
//! themselves. Allocation prefers a freed page over extending the file; the
//! trunk chain is walked with a cycle guard bounded by the file size so a
//! corrupted chain is rejected rather than looped.
//!
//! Trunk layout (big-endian, one page):
//!   `[0:4]`  next trunk page number (0 = end of chain)
//!   `[4:8]`  count of leaf page numbers stored here
//!   `[8:]`   `count` leaf page numbers, 4 bytes each

use crate::consts::{
    HDR_FIRST_FREELIST_OFFSET, HDR_FREELIST_COUNT_OFFSET, PAGE_CRC_COVERED,
};
use crate::error::{Result, StoreError};
use crate::pager::Pager;

const TRUNK_FIXED_SIZE: usize = 8;
const TRUNK_ENTRY_SIZE: usize = 4;

/// Decoded contents of a trunk page.
struct TrunkPage {
    next_trunk: u32,
    leaf_pages: Vec<u32>,
}

impl TrunkPage {
    fn decode(data: &[u8]) -> Result<TrunkPage> {
        let next_trunk = u32::from_be_bytes(data[0..4].try_into().unwrap());
        let leaf_count = u32::from_be_bytes(data[4..8].try_into().unwrap()) as usize;
        let cap = max_leaves(data.len());
        if leaf_count > cap {
            return Err(StoreError::FreelistCorruption {
                detail: format!("trunk leaf_count {leaf_count} exceeds capacity {cap}"),
            });
        }
        let mut leaf_pages = Vec::with_capacity(leaf_count);
        for i in 0..leaf_count {
            let off = TRUNK_FIXED_SIZE + i * TRUNK_ENTRY_SIZE;
            leaf_pages.push(u32::from_be_bytes(data[off..off + 4].try_into().unwrap()));
        }
        Ok(TrunkPage {
            next_trunk,
            leaf_pages,
        })
    }

    fn encode(&self, page_size: usize) -> Vec<u8> {
        let mut buf = vec![0u8; page_size];
        buf[0..4].copy_from_slice(&self.next_trunk.to_be_bytes());
        buf[4..8].copy_from_slice(&(self.leaf_pages.len() as u32).to_be_bytes());
        for (i, pno) in self.leaf_pages.iter().enumerate() {
            let off = TRUNK_FIXED_SIZE + i * TRUNK_ENTRY_SIZE;
            buf[off..off + 4].copy_from_slice(&pno.to_be_bytes());
        }
        buf
    }
}

/// Maximum leaf entries a trunk page can hold without colliding with the
/// per-page checksum slot.
fn max_leaves(page_size: usize) -> usize {
    (PAGE_CRC_COVERED.min(page_size) - TRUNK_FIXED_SIZE) / TRUNK_ENTRY_SIZE
}

/// Allocate one page number, preferring the freelist over file extension.
///
/// When the freelist is empty the file is extended. Otherwise a leaf page is
/// popped from the head trunk; when the head trunk's leaf list is empty the
/// trunk page itself is recycled and the chain advances.
pub fn alloc_page(pager: &Pager) -> Result<u32> {
    let first_trunk = pager.first_freelist_trunk()?;
    if first_trunk == 0 {
        return pager.extend_file();
    }

    let trunk_data = pager.read_page(first_trunk)?;
    let mut trunk = TrunkPage::decode(&trunk_data)?;

    if let Some(allocated) = trunk.leaf_pages.pop() {
        pager.write_page(first_trunk, &trunk.encode(pager.page_size()))?;
        let count = pager.freelist_count()?;
        pager.update_header_u32(HDR_FREELIST_COUNT_OFFSET, count.saturating_sub(1))?;
        Ok(allocated)
    } else {
        // The head trunk holds no leaves; recycle the trunk page itself.
        pager.update_header_u32(HDR_FIRST_FREELIST_OFFSET, trunk.next_trunk)?;
        let count = pager.freelist_count()?;
        pager.update_header_u32(HDR_FREELIST_COUNT_OFFSET, count.saturating_sub(1))?;
        Ok(first_trunk)
    }
}

/// Push `page_no` onto the freelist.
///
/// The page is appended to the head trunk's leaf list. When the head trunk is
/// full (or none exists) `page_no` becomes a new trunk pointing at the old head.
pub fn free_page(pager: &Pager, page_no: u32) -> Result<()> {
    let first_trunk = pager.first_freelist_trunk()?;
    let page_size = pager.page_size();
    let cap = max_leaves(page_size);

    if first_trunk == 0 {
        let new_trunk = TrunkPage {
            next_trunk: 0,
            leaf_pages: Vec::new(),
        };
        pager.write_page(page_no, &new_trunk.encode(page_size))?;
        pager.update_header_u32(HDR_FIRST_FREELIST_OFFSET, page_no)?;
        let count = pager.freelist_count()?;
        pager.update_header_u32(HDR_FREELIST_COUNT_OFFSET, count + 1)?;
        return Ok(());
    }

    let trunk_data = pager.read_page(first_trunk)?;
    let mut trunk = TrunkPage::decode(&trunk_data)?;

    if trunk.leaf_pages.len() < cap {
        trunk.leaf_pages.push(page_no);
        pager.write_page(first_trunk, &trunk.encode(page_size))?;
        let count = pager.freelist_count()?;
        pager.update_header_u32(HDR_FREELIST_COUNT_OFFSET, count + 1)?;
    } else {
        let new_trunk = TrunkPage {
            next_trunk: first_trunk,
            leaf_pages: Vec::new(),
        };
        pager.write_page(page_no, &new_trunk.encode(page_size))?;
        pager.update_header_u32(HDR_FIRST_FREELIST_OFFSET, page_no)?;
        let count = pager.freelist_count()?;
        pager.update_header_u32(HDR_FREELIST_COUNT_OFFSET, count + 1)?;
    }
    Ok(())
}

/// Collect every page number currently on the freelist chain.
///
/// Walks the trunk chain from the header's first-trunk pointer, collecting each
/// trunk page and each leaf page. The walk is bounded by the file size: a
/// repeated page or a page number beyond the file extent is rejected as
/// [`StoreError::FreelistCorruption`].
pub fn freelist_pages(pager: &Pager) -> Result<std::collections::HashSet<u32>> {
    let mut result = std::collections::HashSet::new();
    let db_size = pager.db_size()?;
    let mut trunk_no = pager.first_freelist_trunk()?;
    let mut seen = std::collections::HashSet::new();

    while trunk_no != 0 {
        if !seen.insert(trunk_no) {
            return Err(StoreError::FreelistCorruption {
                detail: format!("chain cycle: page {trunk_no} repeated"),
            });
        }
        if trunk_no > db_size {
            return Err(StoreError::FreelistCorruption {
                detail: format!("trunk page {trunk_no} exceeds db_size={db_size}"),
            });
        }
        result.insert(trunk_no);

        let trunk_data = pager.read_page(trunk_no)?;
        let trunk = TrunkPage::decode(&trunk_data)?;
        for leaf in &trunk.leaf_pages {
            if !seen.insert(*leaf) {
                return Err(StoreError::FreelistCorruption {
                    detail: format!("chain cycle: leaf page {leaf} repeated"),
                });
            }
            result.insert(*leaf);
        }
        trunk_no = trunk.next_trunk;
    }
    Ok(result)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::PathBuf;

    fn temp_db() -> (tempfile::TempDir, PathBuf) {
        let dir = tempfile::tempdir().unwrap();
        let path = dir.path().join("brain.lilli");
        (dir, path)
    }

    #[test]
    fn freelist_empty_alloc_extends_file() {
        let (_d, path) = temp_db();
        let p = Pager::open(&path).unwrap();
        let before = p.db_size().unwrap();
        let pno = alloc_page(&p).unwrap();
        assert_eq!(pno, before + 1);
        assert_eq!(p.db_size().unwrap(), before + 1);
    }

    #[test]
    fn freelist_free_then_alloc_reuses_page_without_growth() {
        let (_d, path) = temp_db();
        let p = Pager::open(&path).unwrap();
        // Allocate a page by extending, then free it.
        let pno = alloc_page(&p).unwrap();
        let size_after_alloc = p.db_size().unwrap();
        free_page(&p, pno).unwrap();
        assert_eq!(p.freelist_count().unwrap(), 1);

        // Re-allocating must return the freed page and not grow the file.
        let reused = alloc_page(&p).unwrap();
        assert_eq!(reused, pno, "freelist must be preferred over extend");
        assert_eq!(p.db_size().unwrap(), size_after_alloc);
        assert_eq!(p.freelist_count().unwrap(), 0);
    }

    #[test]
    fn freelist_three_free_three_alloc_no_leak() {
        let (_d, path) = temp_db();
        let p = Pager::open(&path).unwrap();
        let a = alloc_page(&p).unwrap();
        let b = alloc_page(&p).unwrap();
        let c = alloc_page(&p).unwrap();
        let stable = p.db_size().unwrap();

        free_page(&p, a).unwrap();
        free_page(&p, b).unwrap();
        free_page(&p, c).unwrap();
        assert_eq!(p.freelist_count().unwrap(), 3);

        let mut got = std::collections::HashSet::new();
        got.insert(alloc_page(&p).unwrap());
        got.insert(alloc_page(&p).unwrap());
        got.insert(alloc_page(&p).unwrap());

        // All three freed pages came back, none leaked, file did not grow.
        let mut want = std::collections::HashSet::new();
        want.insert(a);
        want.insert(b);
        want.insert(c);
        assert_eq!(got, want);
        assert_eq!(p.db_size().unwrap(), stable);
        assert_eq!(p.freelist_count().unwrap(), 0);
    }

    #[test]
    fn freelist_cycle_is_detected() {
        let (_d, path) = temp_db();
        let p = Pager::open(&path).unwrap();
        // Build a self-referential trunk: a page whose next_trunk points to
        // itself, set as the head of the chain.
        let trunk_no = alloc_page(&p).unwrap();
        let self_cycle = TrunkPage {
            next_trunk: trunk_no,
            leaf_pages: Vec::new(),
        };
        p.write_page(trunk_no, &self_cycle.encode(p.page_size()))
            .unwrap();
        p.update_header_u32(HDR_FIRST_FREELIST_OFFSET, trunk_no)
            .unwrap();

        let err = freelist_pages(&p).unwrap_err();
        assert!(matches!(err, StoreError::FreelistCorruption { .. }));
    }

    #[test]
    fn trunk_decode_rejects_out_of_range_leaf_count() {
        let (_d, path) = temp_db();
        let p = Pager::open(&path).unwrap();
        // Stamp a trunk page whose declared leaf_count exceeds the page capacity.
        // A naive decoder slice-indexes past the buffer and panics; decode must
        // surface a typed FreelistCorruption instead.
        let trunk_no = alloc_page(&p).unwrap();
        let mut page = vec![0u8; p.page_size()];
        let bogus = (max_leaves(p.page_size()) + 1) as u32;
        page[4..8].copy_from_slice(&bogus.to_be_bytes());
        p.write_page(trunk_no, &page).unwrap();
        p.update_header_u32(HDR_FIRST_FREELIST_OFFSET, trunk_no)
            .unwrap();

        let err = freelist_pages(&p).unwrap_err();
        assert!(matches!(err, StoreError::FreelistCorruption { .. }));
    }
}

File: src/iai_mcp/lillibrain/freelist.py
"""Freelist trunk page, page allocation, and page freeing for the LilliBrain engine.

Freelist trunk pages are journaled as before-images before any allocation or
free operation mutates them, ensuring that an aborted transaction fully restores
the freelist state (no page leaks, no double-frees).

Invariants enforced here:
  Trunk page before-image recorded before any alloc mutates the trunk.
  Trunk page before-image recorded before free_page mutates the trunk.
  DB header page (page 1) before-image recorded alongside trunk on the alloc path.
"""
from __future__ import annotations

import struct
from dataclasses import dataclass, field
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from iai_mcp.lillibrain.pager import Journal, Pager

# ---------------------------------------------------------------------------
# Trunk page binary layout (big-endian, one page)
# ---------------------------------------------------------------------------
#
#  Offset  Size  Description
#  ------  ----  -----------
#   0       4    next_trunk  (0 = end of chain)
#   4       4    leaf_count  (number of free leaf page numbers stored here)
#   8+      4*N  leaf_page_no[0..N-1]
#
# Max leaf entries per trunk page:
#   (page_size - 8) // 4
#
_TRUNK_FIXED_SIZE = 8       # next_trunk(4) + leaf_count(4)
_TRUNK_ENTRY_SIZE = 4       # each leaf page number


# ---------------------------------------------------------------------------
# Data container
# ---------------------------------------------------------------------------


@dataclass
class FreelistTrunkPage:
    """Decoded content of a freelist trunk page.

    next_trunk: page number of the next trunk (0 = end of chain).
    leaf_pages: list of free leaf page numbers stored in this trunk.
    """

    next_trunk: int
    leaf_pages: list[int] = field(default_factory=list)

    @classmethod
    def decode(cls, data: bytes | bytearray | memoryview) -> FreelistTrunkPage:
        """Decode a raw page buffer into a FreelistTrunkPage.

        Reads next_trunk and leaf_count from the fixed header, then reads
        leaf_count page numbers from the variable-length body.
        """
        (next_trunk,) = struct.unpack_from(">I", data, 0)
        (leaf_count,) = struct.unpack_from(">I", data, 4)
        leaf_pages: list[int] = []
        for i in range(leaf_count):
            offset = _TRUNK_FIXED_SIZE + i * _TRUNK_ENTRY_SIZE
            (pno,) = struct.unpack_from(">I", data, offset)
            leaf_pages.append(pno)
        return cls(next_trunk=next_trunk, leaf_pages=leaf_pages)

    def encode(self, page_size: int) -> bytes:
        """Encode this trunk page into a byte string of exactly page_size bytes.

        Raises ValueError if the leaf_pages list exceeds the maximum capacity
        for the given page size.
        """
        max_leaves = (page_size - _TRUNK_FIXED_SIZE) // _TRUNK_ENTRY_SIZE
        if len(self.leaf_pages) > max_leaves:
            raise ValueError(
                f"FreelistTrunkPage: {len(self.leaf_pages)} leaf entries exceed "
                f"max {max_leaves} for page_size={page_size}"
            )
        buf = bytearray(page_size)
        struct.pack_into(">I", buf, 0, self.next_trunk)
        struct.pack_into(">I", buf, 4, len(self.leaf_pages))
        for i, pno in enumerate(self.leaf_pages):
            struct.pack_into(">I", buf, _TRUNK_FIXED_SIZE + i * _TRUNK_ENTRY_SIZE, pno)
        return bytes(buf)


# ---------------------------------------------------------------------------
# Corruption error
# ---------------------------------------------------------------------------


class FreelistCorruptionError(RuntimeError):
    """Freelist trunk chain is internally inconsistent (cycle or invalid page number)."""


# ---------------------------------------------------------------------------
# Freelist alloc/free — trunk-journaling invariants
# ---------------------------------------------------------------------------


def alloc_page(pager: Pager, *, journal: Journal | None) -> int:
    """Allocate one page number, preferring the freelist over file extension.

    In rollback-journal mode, the freelist trunk page and DB header page
    (page 1) are recorded as before-images before any in-memory modification,
    so that an aborted transaction can fully restore the freelist state.
    In write-ahead mode, journal is None and before-image recording is skipped
    (WAL rollback rewinds via the checksum-chain snapshot, not before-images).

    Returns the allocated page number (1-indexed).
    """
    first_trunk = pager.read_db_header_field("first_freelist_trunk")

    if first_trunk == 0:
        # No freelist — extend the file (no trunk to journal)
        return pager.extend_file()

    # Journal the trunk page before-image before any mutation (rollback mode only).
    trunk_data = pager.read_page(first_trunk)
    if journal is not None:
        journal.record_before_image(first_trunk, trunk_data)

    # Journal the DB header page (page 1 holds freelist_count and
    # first_freelist_trunk) before any freelist mutation (rollback mode only).
    if journal is not None:
        journal.record_before_image(1, pager.read_page(1))

    trunk = FreelistTrunkPage.decode(trunk_data)

    if trunk.leaf_pages:
        # Pop a leaf page from the trunk
        allocated = trunk.leaf_pages.pop()
        pager.write_page(first_trunk, trunk.encode(pager._page_size))
        freelist_count = pager.read_db_header_field("freelist_count")
        pager.update_db_header("freelist_count", max(0, freelist_count - 1))
        return allocated
    else:
        # Trunk leaf list is empty — pop the trunk page itself, advance chain
        next_trunk = trunk.next_trunk
        pager.update_db_header("first_freelist_trunk", next_trunk)
        freelist_count = pager.read_db_header_field("freelist_count")
        pager.update_db_header("freelist_count", max(0, freelist_count - 1))
        return first_trunk


def free_page(pager: Pager, page_no: int, *, journal: Journal | None) -> None:
    """Push page_no onto the freelist trunk and increment freelist_count.

    In rollback-journal mode, the current trunk page is recorded as a
    before-image before any mutation so that an aborted transaction can restore
    the trunk state.  In write-ahead mode, journal is None and before-image
    recording is skipped (WAL rollback rewinds via the checksum-chain snapshot).

    If the current trunk page is full or no trunk exists, a new trunk page
    is allocated to hold the freed page.
    """
    first_trunk = pager.read_db_header_field("first_freelist_trunk")
    page_size = pager._page_size
    max_leaves = (page_size - _TRUNK_FIXED_SIZE) // _TRUNK_ENTRY_SIZE

    # Journal the DB header page (page 1) before any mutation (rollback mode only).
    header_data = pager.read_page(1)
    if journal is not None:
        journal.record_before_image(1, header_data)

    if first_trunk == 0:
        # No current trunk — page_no itself becomes the new trunk (with no leaves)
        new_trunk = FreelistTrunkPage(next_trunk=0, leaf_pages=[])
        pager.write_page(page_no, new_trunk.encode(page_size))
        pager.update_db_header("first_freelist_trunk", page_no)
        freelist_count = pager.read_db_header_field("freelist_count")
        pager.update_db_header("freelist_count", freelist_count + 1)
        return

    # Load the current trunk and check capacity
    trunk_data = pager.read_page(first_trunk)
    trunk = FreelistTrunkPage.decode(trunk_data)

    if len(trunk.leaf_pages) < max_leaves:
        # Journal the trunk before-image before mutation (rollback mode only).
        if journal is not None:
            journal.record_before_image(first_trunk, trunk_data)
        trunk.leaf_pages.append(page_no)
        pager.write_page(first_trunk, trunk.encode(page_size))
        freelist_count = pager.read_db_header_field("freelist_count")
        pager.update_db_header("freelist_count", freelist_count + 1)
    else:
        # Current trunk is full — page_no becomes the new trunk, points to old trunk
        new_trunk = FreelistTrunkPage(next_trunk=first_trunk, leaf_pages=[])
        pager.write_page(page_no, new_trunk.encode(page_size))
        pager.update_db_header("first_freelist_trunk", page_no)
        freelist_count = pager.read_db_header_field("freelist_count")
        pager.update_db_header("freelist_count", freelist_count + 1)


def freelist_pages(pager: Pager) -> set[int]:
    """Return the complete set of page numbers currently on the freelist chain.

    Walks the trunk chain from first_freelist_trunk, collecting every trunk
    page number and every leaf page number. Guards against cycles by bounding
    the walk by db_size — raises FreelistCorruptionError if a page repeats.
    """
    result: set[int] = set()
    db_size = pager.read_db_header_field("db_size")
    trunk_no = pager.read_db_header_field("first_freelist_trunk")
    seen: set[int] = set()

    while trunk_no != 0:
        if trunk_no in seen:
            raise FreelistCorruptionError(
                f"Freelist chain cycle detected: page {trunk_no} repeated "
                f"(seen={sorted(seen)})"
            )
        if trunk_no > db_size:
            raise FreelistCorruptionError(
                f"Freelist trunk page {trunk_no} exceeds db_size={db_size}"
            )
        seen.add(trunk_no)
        result.add(trunk_no)

        trunk_data = pager.read_page(trunk_no)
        trunk = FreelistTrunkPage.decode(trunk_data)

        for leaf_no in trunk.leaf_pages:
            if leaf_no in seen:
                raise FreelistCorruptionError(
                    f"Freelist chain cycle detected: leaf page {leaf_no} repeated "
                    f"(seen={sorted(seen)})"
                )
            seen.add(leaf_no)
            result.add(leaf_no)

        trunk_no = trunk.next_trunk

    return result
